File: sorts.cpp
#include <iostream>
#include <vector>

using namespace std;

#define MIN(a,b)((a)<(b)?(a):(b))

void swap(int &a, int &b){
    int temp = a;
    a = b;
    b = temp;
}

//Bubble Sort
int bubbleSort(vector<int> &array){
    int executions = 0;
    for(int i = 0; i < array.size(); i++){
        for(int j = 0; j < array.size(); j++){
            executions++;
            if(array[i] < array[j]){
                swap(array[i], array[j]);
            }
        }
    }
    return executions;
}

//Insertion Sort
int insertionSort(vector<int> &array){
    int valueInsert = 0, hole = 0, executions = 0;
    for(int i = 1; i < array.size(); i++){
        valueInsert = array[i];
        hole = i;
        while(valueInsert < array[hole - 1] && hole > 0){
            array[hole] = array[hole - 1];
            hole--;
            executions++;
        }
        executions++;
        array[hole] = valueInsert;
    }
    return executions;
}

//Selection Sort
int selectionSort(vector<int> &array){
    int executions = 0;
    for(int i = 0; i < array.size(); i++){
        int min = array[i], minLoc = i;
        for(int j = i; j < array.size(); j++){
            executions++;
            if(MIN(min, array[j]) != min){
                minLoc = j;
                min = array[j];
            }
        }
        swap(array[i], array[minLoc]);
    }
    return executions;
}

//Merge Sort
// void mergeSort(vector<int> &array, int l, int r){
//     if(l < r){
//         int m = l+(r-1)/2;
//         mergeSort(array, l, m); 
//         mergeSort(array, m+1, r); 

//         //Merge
//         int i, j, k; 
//         int n1 = m - l + 1; 
//         int n2 =  r - m; 
    
//         /* create temp arrays */
//         vector<int> L, R;
//         L.resize(n1);
//         R.resize(n2); 
    
//         /* Copy data to temp arrays L[] and R[] */
//         for (i = 0; i < n1; i++) 
//             L[i] = array[l + i]; 
//         for (j = 0; j < n2; j++) 
//             R[j] = array[m + 1+ j]; 
    
//         /* Merge the temp arrays back into arr[l..r]*/
//         i = 0; // Initial index of first subarray 
//         j = 0; // Initial index of second subarray 
//         k = l; // Initial index of merged subarray 
//         while (i < n1 && j < n2) 
//         { 
//             if (L[i] <= R[j]) 
//             { 
//                 array[k] = L[i]; 
//                 i++; 
//             } 
//             else
//             { 
//                 array[k] = R[j]; 
//                 j++; 
//             } 
//             k++; 
//         } 
    
//         /* Copy the remaining elements of L[], if there 
//         are any */
//         while (i < n1) 
//         { 
//             array[k] = L[i]; 
//             i++; 
//             k++; 
//         } 
    
//         /* Copy the remaining elements of R[], if there 
//         are any */
//         while (j < n2) 
//         { 
//             array[k] = R[j]; 
//             j++; 
//             k++; 
//         } 
//     }
// }


// void Sort(vector<int> &array){

// }

int main(int argc, char* argv[]){
    vector<int> array = {10, 4, 25, 50, 75, 22, 21, 2, 6, 5, 3};

    // mergeSort(array, 0, array.size() - 1);
    cout << "num of exections(n = " << array.size() << "): " << insertionSort(array) << endl;
    for(int i = 0; i < array.size(); i++)
        cout << array[i] << endl;
    
    return 0;
}

File: sorts.py
import time
import random

def swap(arr, a, b):
    temp = arr[a]
    arr[a] = arr[b]
    arr[b] = temp

#O(N^2)
def bubbleSort(arr):
    for i in range(len(arr)):
        for j in range(len(arr)):
            if arr[i] <= arr[j]:
                swap(arr, i, j)
    return arr

#O(N^2)
def selectionSort(arr):
    for i in range(len(arr)):
        min = i
        for j in range(i, len(arr)):
            if arr[min] > arr[j]:
                min = j
        swap(arr, i, min)
    return arr

#O(N^2)
def insertionSort(arr):
    sorted = []
    for i in range(len(arr)):
        if i is 0:
            sorted.append(arr[i])
        elif(sorted[i-1] <= arr[i]):
            sorted.append(arr[i])
        else:
            for j in range(len(sorted)):
                if sorted[j] > arr[i]:
                    sorted.insert(j, arr[i])
                    break
    return sorted

#Helper function for MergeSort
def merge(n1, n2):
    a = 0
    b = 0
    arr = []
    for i in range(len(n1) + len(n2)):
        if(b >= len(n2) or (a < len(n1) and n1[a] < n2[b])):
            arr.append(n1[a])
            a += 1
        else:
            arr.append(n2[b])
            b +=1
    return arr

#O(N log(N))
def mergeSort(arr):
    if len(arr) is 1:
        return arr
    middle = len(arr)//2
    left = arr[:middle]
    right = arr[middle:]
    result = merge(mergeSort(left), mergeSort(right))
    return result
    
def test1():
    array = [5,4,3,2,1,2,3,4,5, 8, 9, 6, 1, 10]
    print(mergeSort(array))
    print(bubbleSort(array))
    print(selectionSort(array))
    print(insertionSort(array))

def test2():
    array = []
    elements = 20
    for i in range(elements):
        array.append(int(random.random() * 1000000))
    print("array: ", array)
    print(mergeSort(array))
    print(bubbleSort(array))
    print(selectionSort(array))
    print(insertionSort(array))
    array.sort()
    print(array)

def timeTrial():
    array = []
    elements = 10000
    for i in range(elements):
        array.append(random.random() * 1000000)
    
    #Bubble Sort
    t1 = time.time()
    bubbleSort(array)
    t2 = time.time()
    print("this is how much time it takes to do bubbleSort: ", (t2-t1))

    #Selection Sort
    t1 = time.time()
    selectionSort(array)
    t2 = time.time()
    print("this is how much time it takes to do selectionSort: ", (t2-t1))

    #Insertion Sort
    t1 = time.time()
    insertionSort(array)
    t2 = time.time()
    print("this is how much time it takes to do insertionSort: ", (t2-t1))

    #Merge Sort
    t1 = time.time()
    mergeSort(array)
    t2 = time.time()
    print("this is how much time it takes to do mergeSort: ", (t2-t1))

    #Merge Sort
    t1 = time.time()
    array.sort()
    t2 = time.time()
    print("this is how much time it takes to do provided sort: ", (t2-t1))


test1()
test2()
timeTrial()